bullet/src/EntityManagementFeatures.cc

```diff
@@ -0,0 +1,213 @@
+  const auto dispatcher =
+    std::make_shared<btCollisionDispatcher>(collisionConfiguration.get());
+  //const auto broadphase = std::shared_ptr<btBroadphaseInterface>(new btDbvtBroadphase());
+  const auto broadphase = std::make_shared<btDbvtBroadphase>();
+  const auto solver =
+    std::make_shared<btSequentialImpulseConstraintSolver>();
+  const auto world = std::make_shared<btDiscreteDynamicsWorld>(
+    dispatcher.get(), broadphase.get(), solver.get(), collisionConfiguration.get());
+
+  /* TO-DO(Lobotuerk): figure out what this line does*/
+  world->getSolverInfo().m_globalCfm = 0;
+
+  btGImpactCollisionAlgorithm::registerAlgorithm(dispatcher.get());
+
+  return this->AddWorld(
+    {_name, collisionConfiguration, dispatcher, broadphase, solver, world});
+}
+
+/////////////////////////////////////////////////
+bool EntityManagementFeatures::RemoveModel(const Identity &_modelID)
+{
+  // Check if the model exists
+  if (this->models.find(_modelID.id) == this->models.end())
+  {
+    return false;
+  }
+
+  auto model = this->models.at(_modelID);
+  auto bulletWorld = this->worlds.at(model->world)->world;
+
+  // Clean up joints, this section considers both links in the joint
+  // are part of the same world
+  auto joint_it = this->joints.begin();
+  while (joint_it != this->joints.end())
+  {
+    const auto &jointInfo = joint_it->second;
+    const auto &childLinkInfo = this->links[jointInfo->childLinkId];
+    if (childLinkInfo->model.id == _modelID.id)
+    {
+      bulletWorld->removeConstraint(jointInfo->joint.get());
+      joint_it = this->joints.erase(joint_it);
+      continue;
+    }
+    joint_it++;
+  }
+
+  // Clean up collisions
+  auto collision_it = this->collisions.begin();
+  while (collision_it != this->collisions.end())
+  {
+    const auto &collisionInfo = collision_it->second;
+    if (collisionInfo->model.id == _modelID.id)
+    {
+      collision_it = this->collisions.erase(collision_it);
+      continue;
+    }
+    collision_it++;
+  }
+
+  // Clean up links
+  auto it = this->links.begin();
+  while (it != this->links.end())
+  {
+    const auto &linkInfo = it->second;
+    if (linkInfo->model.id == _modelID.id)
+    {
+      bulletWorld->removeRigidBody(linkInfo->link.get());
+      it = this->links.erase(it);
+      continue;
+    }
+    it++;
+  }
+
+  // Clean up model, links are erased and the model is just a name to tie
+  // them together
+  this->models.erase(_modelID.id);
+
+  return true;
+}
+
+bool EntityManagementFeatures::ModelRemoved(
+    const Identity &_modelID) const
+{
+  return this->models.find(_modelID) == this->models.end();
+}
+
+bool EntityManagementFeatures::RemoveModelByIndex(
+    const Identity & _worldID, std::size_t _modelIndex)
+{
+  // Check if the model exists
+  if (this->models.find(_modelIndex) == this->models.end() ||
+      this->models.at(_modelIndex)->world.id != _worldID.id)
+  {
+    return false;
+  }
+
+  auto model = this->models.at(_modelIndex);
```

Comment:
I think this can be refactored into it's own function because `RemoveModel` is doing the same thing.

Reply:
You are right, fff1f08. This also considers index <-> ID.id